ql/src/java/org/apache/hadoop/hive/ql/Context.java

```diff
@@ -625,17 +637,43 @@ private Path getExternalScratchDir(URI extURI) {
     return getStagingDir(new Path(extURI.getScheme(), extURI.getAuthority(), extURI.getPath()), !isExplainSkipExecution());
   }
 
+  /**
+   * Remove any created scratch directories.
+   */
+  public void removeResultCacheDir() {
+    if(this.fsResultCacheDirs != null) {
+      try {
+        Path p = this.fsResultCacheDirs;
+        FileSystem fs = p.getFileSystem(conf);
+        LOG.debug("Deleting result cache dir: {}", p);
+        fs.delete(p, true);
+        fs.cancelDeleteOnExit(p);
+      } catch (Exception e) {
+        LOG.warn("Error Removing result cache dir: "
+                     + StringUtils.stringifyException(e));
+      }
+    }
+  }
+
   /**
    * Remove any created scratch directories.
    */
   public void removeScratchDir() {
+    String resultCacheDir = null;
+    if(this.fsResultCacheDirs != null) {
+      resultCacheDir = this.fsResultCacheDirs.toUri().getPath();
+    }
     for (Map.Entry<String, Path> entry : fsScratchDirs.entrySet()) {
       try {
         Path p = entry.getValue();
+        if(resultCacheDir == null || !p.toUri().getPath().contains(resultCacheDir)) {
```

Comment:
One example which I believe I have seen:

fsScratchDirs = [ /base/scratchdir/path ]
resultCacheDir = [ /base/scratchdir/path/with/resultcache/ ]

The result of this logic change will be that /base/scratchdir/path/ goes undeleted. With resultCacheDir eventually being deleted later. One question, is there any additional data in /base/scratchdir/path/ besides just the final query result, such as other temporary data generated by the query? If so, this additional data, in addition to the /base/scratchdir/path/ directory itself (empty or not) will be left around.